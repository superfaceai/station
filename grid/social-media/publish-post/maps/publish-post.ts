/* eslint-disable jest/no-export */
import {
  RecordingProcessOptions,
  SuperfaceTest,
  SuperfaceTestRun,
} from '@superfaceai/testing';

type ProfilesResult = Array<{ id: string; name: string }>;

export const getPublishingProfiles = async (
  provider: string
): Promise<ProfilesResult> => {
  const superfacePublishingProfiles = new SuperfaceTest({
    profile: 'social-media/publishing-profiles',
    provider,
  });
  const result = await superfacePublishingProfiles.run({
    useCase: 'GetProfilesForPublishing',
    input: {},
  });
  expect(result.isOk()).toBeTruthy();

  return (result.unwrap() as { profiles: ProfilesResult })?.profiles || [];
};

export const publishPostTest = (
  provider: string,
  hooks?: RecordingProcessOptions
): void => {
  describe(`social-media/publish-post/${provider}`, () => {
    let superfacePublishingProfiles: SuperfaceTest;
    let superfacePublisPost: SuperfaceTest;

    beforeEach(() => {
      superfacePublishingProfiles = new SuperfaceTest({
        profile: 'social-media/publishing-profiles',
        provider,
      });
      superfacePublisPost = new SuperfaceTest({
        profile: 'social-media/publish-post',
        provider,
      });
    });

    describe('PublishPost', () => {
      describe('when publishing text post', () => {
        it('should succeed', async () => {
          const result = await superfacePublishingProfiles.run({
            useCase: 'GetProfilesForPublishing',
            input: {},
          });

          expect(result.isOk()).toBeTruthy();
          const resultUnwrapped = result.unwrap();

          await expect(
            superfacePublisPost.run(
              {
                useCase: 'PublishPost',
                input: {
                  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
                  profileId: (resultUnwrapped as any).profiles[0].id,
                  text: `Test from Superface Station 7.`,
                },
              },
              hooks
            )
          ).resolves.toMatchSnapshot();
        });
      });
    });
  });
};

type ErrorCase = {
  name: string;
  input: SuperfaceTestRun['input'];
};

export const publishPostErrorTest = (
  provider: string,
  cases: ErrorCase[],
  hooks?: RecordingProcessOptions
): void => {
  describe(`social-media/publish-post/${provider}`, () => {
    let superfacePublisPost: SuperfaceTest;

    beforeEach(() => {
      superfacePublisPost = new SuperfaceTest({
        profile: 'social-media/publish-post',
        provider,
      });
    });

    const testCases: Array<
      [string, ErrorCase['input']]
    > = cases.map(({ name, input }) => [name, input]);

    describe('PublishPost', () => {
      describe('error cases', () => {
        // FIXME: Use $name in Jest 27+
        test.each(testCases)('%s', async (_name, input) => {
          const result = await superfacePublisPost.run(
            {
              useCase: 'PublishPost',
              input,
            },
            hooks
          );
          expect(result.isErr()).toBe(true);
          expect(result).toMatchSnapshot();
        });
      });
    });
  });
};
