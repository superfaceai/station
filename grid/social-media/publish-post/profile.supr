"""
Publish Social Media Post
Publish posts on social media such as Instagram, Facebook, or Twitter.
"""

name = "social-media/publish-post"
version = "1.2.0"

"""
Publish a Post
Publish a post on a social media profile with optional image attachments.
"""
usecase PublishPost unsafe {
  input {
    """
    Profile ID
    Identifier of a profile to publish to. May be optional with some providers.
    """
    profileId string!

    """
    Text
    Text of the post
    """
    text string!

    """
    Link
    URL to attach to the post
    """
    link string!

    """
    Title
    Title of the post, if supported by the provider (e.g. Pinterest). Ignored otherwise.
    """
    title string!

    """
    Media attachments
    Only images and gifs are supported
    """
    media [
      MediaItem
    ]

    attachments [
      MediaAttachment
    ]
  }

  result {

    """
    Post ID
    Identifier of the published post.
    """
    postId string

    """
    Post URL
    Link to the published post.
    """
    url string
  }

  error {
    """
    Title
    A short, human-readable summary of the problem type.
    """
    title! string

    """
    Detail
    A human-readable explanation specific to this occurrence of the problem.
    """
    detail string

    """
    Code
    Normalized error code usable across providers
    """
    code ErrorCode

    """
    Original Error
    Original response from the provider
    """
    original
  }

  example Successful {
    input {
      profileId = '110658944785122',
      text = 'Text of the Post.',
    }

    result {
      postId = '110658944785122_125682506616099',
      url = 'https://www.facebook.com/110658944785122/posts/125682506616099/',
    }
  }

  example Failed {
    input {
      profileId = 'invalid profile ID',
      text = 'Text of the Post.',
    }

    error {
      title = 'Bad request'
    }
  }
}

model MediaAttachment {
  """
  Media Container ID
  ID of a previously uploaded media item, e.g. a video.
  """
  id! string!
}

model MediaItem {
  """
  URL of image
  Will be fetched by provider, must be accessible publicly. 
  If both `contents` and `url` fields are provided, a preferable option will be picked by the map.
  """
  url string!
  """
  Buffer or base64 encoded string with image
  If both `contents` and `url` fields are provided, a preferable option will be picked by the map.
  """
  contents
  "Alternative text for screen readers"
  altText string
}

// https://developers.facebook.com/docs/instagram-api/reference/error-codes/
// https://developers.facebook.com/docs/graph-api/guides/error-handling/
// https://developers.facebook.com/docs/graph-api/overview/rate-limiting#error-codes

model ErrorCode enum {
  """
  Bad request
  The request was invalid or cannot be otherwise served.
  """
  BadRequest
  """
  Not found
  The profileId, container or other input item does not exist.
  """
  NotFound
  """
  Rate limit reached
  Returned when a request cannot be served due to the user or app reached max requests quota.
  """
  RateLimitReached
  """
  Unauthenticated
  There was a problem authenticating your request. Check that you are passing valid access token.
  """
  Unauthenticated
  """
  Unauthorized
  The request is understood, but it has been refused or access is not allowed.
  """
  Unauthorized

  Timeout
  ServerError
  Spam
  RestrictedProfile
  PayloadTooLarge
  UnsupportedAttachmentFormat

  """
  Unknown error
  Returned when an unexpected error occured.
  """
  UnknownError
}
