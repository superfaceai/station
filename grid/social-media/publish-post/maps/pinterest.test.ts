import {
  publishMediaPostTest,
  publishMediaUploadTest,
} from './publish-media-post';
import { publishPostErrorTest } from './publish-post';

publishMediaPostTest('pinterest', {
  title: `Test post title`,
  text: `Test media publishing from Superface Station.`,
  media: [
    {
      url: 'https://upload.wikimedia.org/wikipedia/commons/0/09/Sitta-carolinensis-001.jpg',
    },
  ],
});
publishPostErrorTest('pinterest', [
  {
    name: 'missing media causes immediate failure',
    input: { profileId: '4', text: 'This should fail.' },
  },
  {
    name: 'invalid profileId ends in not found error',
    input: {
      profileId: '4',
      text: `Test media publishing from Superface Station.`,
      media: [
        {
          url: 'https://upload.wikimedia.org/wikipedia/commons/0/09/Sitta-carolinensis-001.jpg',
        },
      ],
    },
  },
]);
publishMediaUploadTest('pinterest', [
  { name: 'JPG upload passes', media: ['horizontal.jpg'] },
  { name: 'transparent PNG upload passes', media: ['vase.transparent.png'] },
  {
    name: 'transparent WebP upload fails',
    media: ['vase.transparent.webp'],
    success: false,
  },
]);
