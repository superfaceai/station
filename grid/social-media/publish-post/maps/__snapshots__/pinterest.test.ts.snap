// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`social-media/publish-post/pinterest PublishPost error cases invalid profileId ends in not found error 1`] = `
Err {
  "error": "HTTPError: Expected HTTP error
Properties: {
  \\"title\\": \\"Not found\\",
  \\"detail\\": \\"Board not found.\\",
  \\"original\\": {
    \\"code\\": 40,
    \\"message\\": \\"Board not found.\\"
  }
}
AST Path: definitions[0].statements[8].responseHandlers[4].statements[0]
Original Map Location: Line 96, column 7",
}
`;

exports[`social-media/publish-post/pinterest PublishPost error cases missing media causes immediate failure 1`] = `
Err {
  "error": "MappedError: Expected error
Properties: {
  \\"title\\": \\"Invalid media\\",
  \\"detail\\": \\"You must provide exactly one media item for Pinterest.\\"
}
AST Path: definitions[0].statements[0]
Original Map Location: Line 6, column 3",
}
`;

exports[`social-media/publish-post/pinterest PublishPost media upload JPG upload passes 1`] = `
Ok {
  "value": Object {
    "postId": "989877193072294310",
    "url": "https://www.pinterest.com/pin/989877193072294310/",
  },
}
`;

exports[`social-media/publish-post/pinterest PublishPost media upload transparent PNG upload passes 1`] = `
Ok {
  "value": Object {
    "postId": "989877193072294312",
    "url": "https://www.pinterest.com/pin/989877193072294312/",
  },
}
`;

exports[`social-media/publish-post/pinterest PublishPost media upload transparent WebP upload fails 1`] = `
Err {
  "error": "MappedError: Expected error
Properties: {
  \\"title\\": \\"Unknown or unsupported media contents type\\",
  \\"detail\\": \\"Only PNG or JPG images are supported.\\"
}
AST Path: definitions[0].statements[7]
Original Map Location: Line 43, column 3",
}
`;

exports[`social-media/publish-post/pinterest PublishPost when publishing media post should succeed 1`] = `
Ok {
  "value": Object {
    "postId": "989877193072294309",
    "url": "https://www.pinterest.com/pin/989877193072294309/",
  },
}
`;
